### ISSUE TITLE
qtum-qt crashes on backup wallet, via menu

### ISSUE THREAD

**Issue by rolf-schenk** [qtumproject/qtum#217]
### Describe the issue
qtum-qt process crashes silently when trying to backup the wallet via menu
Note: backing up the wallet via console (backupwallet) is working fine. 

### Can you reliably reproduce the issue?
1. choose menu "File" --> "Backup wallet" --> process is being killed  with message src/tcmalloc.cc:283] Attempt to free invalid pointer 0x7f9b0ac0cfd0 
Abort trap: 6

### Apple stack infos
Process:               qtum-qt [1337]
Path:                  /Users/USER/Documents/*/qtum-qt
Identifier:            qtum-qt
Version:               0
Code Type:             X86-64 (Native)
Parent Process:        ??? [1]
Responsible:           qtum-qt [1337]
System Integrity Protection: enabled
Crashed Thread:        24  Dispatch queue: com.apple.NSXPCConnection.user.1344
Exception Type:        EXC_CRASH (SIGABRT)
Exception Codes:       0x0000000000000000, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

16  org.qt-project.QtWidgets      	0x000000010ade58ac QDialog::exec() + 284
17  org.qt-project.QtWidgets      	0x000000010ada79fe QFileDialog::getSaveFileUrl(QWidget*, QString const&, QUrl const&, QString const&, QString*, QFlags<QFileDialog::Option>, QStringList const&) + 398
18  org.qt-project.QtWidgets      	0x000000010ada778e QFileDialog::getSaveFileName(QWidget*, QString const&, QString const&, QString const&, QString*, QFlags<QFileDialog::Option>) + 254
19  qtum-qt                       	0x00000001099da000 GUIUtil::getSaveFileName(QWidget*, QString const&, QString const&, QString const&, QString*) + 192 (guiutil.cpp:321)
20  qtum-qt                       	0x0000000109aaee7f WalletView::backupWallet() + 127 (qstring.h:1084)
21  org.qt-project.QtCore         	0x000000010b84eb4a QMetaObject::activate(QObject*, int, int, void**) + 2954
22  org.qt-project.QtWidgets      	0x000000010ab9b665 QAction::activate(QAction::ActionEvent) + 309
23  org.qt-project.QtCore         	0x000000010b84eb4a QMetaObject::activate(QObject*, int, int, void**) + 2954
24  org.qt-project.QtCore         	0x000000010b846de4 QObject::event(QEvent*) + 788
25  org.qt-project.QtWidgets      	0x000000010aba4e82 QApplicationPrivate::notify_helper(QObject*, QEvent*) + 306
26  org.qt-project.QtWidgets      	0x000000010aba619f QApplication::notify(QObject*, QEvent*) + 383
27  org.qt-project.QtCore         	0x000000010b81e6ff QCoreApplication::notifyInternal2(QObject*, QEvent*) + 159
28  org.qt-project.QtCore         	0x000000010b81f792 QCoreApplicationPrivate::sendPostedEvents(QObject*, int, QThreadData*) + 850

### What version of Qtum are you using?
- Qtum Core Version v0.14.0.0-da3357f4d (64-Bit)
- git show
-- commit da3357f4d44436bd8dbd389ca6ddb6f33f485f78
-- Merge: e2d7f5e7b 412fb0292

### Machine specs:
Hardware-Übersicht:
  (osx sierra 10.12.5 (16F73))
  Modell-Identifizierung:	iMac15,1
  Prozessortyp:	Intel Core i7
  Prozessorgeschwindigkeit:	4 GHz
  Anzahl der Prozessoren:	1
  Gesamtanzahl der Kerne:	4
  Speicher:	16 GB
  Disk: 3TB Fusion

**Comment by tekahuna** [comment]
I am on also on Sierra, but I do not appear to have the same issue.  Backing up both encrypted, and unencrypted seems fine.

** What version of Qtum are you using?**
- Qtum Core version v0.14.0.0-f69fb1777 (64-bit)
- git show
-- commit f69fb17
-- Merge: 13ba3907c 217a5bab1
- brew update; brew upgrade @ 06/30/2017
- Built with Qt5 according to Sparknet Release criteria

** Machine Specs: **
Custom Gigabyte GA-Z87X-UD5H based Hackintosh
OS: macOS Sierra 10.12.5 (16F73)
Model:  iMac14,2 (SMBIOS faked via Clover)
Processor Type:  Intel Core i7-4770 Haswell
Processor Speed:  3.5Ghz
Number of Processors:  1
Total Number of Cores:  4
Memory:  32 GB
Disk: 2TB Fusion(System), 512GB Plextor SSD (Development)

**Comment by rolf-schenk** [comment]
I started from scratch with versions:
- brew update; brew upgrade 2/Jul/2017
- also updated git, now using 2.13.2
- Qtum Core Version v0.14.0.0-f69fb1777 (64-Bit)
- git show
-- commit f69fb1777354b465b97f511994085e7eefc062ea (HEAD -> testnet-1, origin/testnet-1, origin/HEAD)
-- Merge: 13ba3907c 217a5bab1 
- qt: stable 5.9.1 (bottled), HEAD [keg-only]  (was on that version before too)
- autogen / configure / make acc. to instructions in /release

Still the qt crashes when backing up the the wallet via menu.

During autogen.sh I notice this error - do you also see that?

`
std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&) in libjsonrpccpp-client.a(httpclient.cpp.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [ethminer/ethminer] Error 1
make[1]: *** [ethminer/CMakeFiles/ethminer.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 70%] Building CXX object libweb3jsonrpc/CMakeFiles/web3jsonrpc.dir/AccountHolder.cpp.o
`

**Comment by tekahuna** [comment]
Your error appears to be a compiler/linker mismatch problem. See this link for more info:
[https://stackoverflow.com/questions/7533321/error-when-compiling-some-simple-c-code](url)

I get warnings for no symbols, but not the failure you get.  Here is what my errors looks like:

`/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: libweb3jsonrpc.a(WinPipeServer.cpp.o) has no symbols`
`/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: libweb3jsonrpc.a(WinPipeServer.cpp.o) has no symbols`

From this warning, it would appear that my working version is built and linked with the Xcode Default toolchain.  Here are my versions:

`$ gcc -v`
`Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1`
`Apple LLVM version 8.1.0 (clang-802.0.42)`
`Target: x86_64-apple-darwin16.6.0`
`Thread model: posix`
`InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin`
`$ clang -v`
`Apple LLVM version 8.1.0 (clang-802.0.42)`
`Target: x86_64-apple-darwin16.6.0`
`Thread model: posix`
`InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin`
`$ ranlib`
`error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: no archives specified`
`Usage: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib [-sactfqLT] [-] archive [...]`

What is your output from `gcc -v`, `clang -v` & `ranlib`?

Also, do you find anything suspicious in `/Your/path/to/qtum/src/cpp-ethereum/deps/src/jsonrpccpp-stamp/jsonrpccpp-configure-*.log` & `/Your/path/to/qtum/src/cpp-ethereum/deps/src/jsonrpccpp-stamp/jsonrpccpp-install-*.log`? I get nothing in my "err" log files.  You?

**Comment by qtum-neil** [comment]
Cannot reproduce this, backup is working fine for me.

qt: stable 5.9.0 (bottled), HEAD [keg-only]
boost: stable 1.64.0 (bottled), HEAD
macOS Sierra: 10.12.5 (16F73)
Apple LLVM version 8.0.0 (clang-800.0.42.1)

**Comment by tekahuna** [comment]
Here is my output from ./configure in regards to gcc, and linker:

> checking whether the gcc linker (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld) supports shared libraries... yes
checking dynamic linker characteristics... darwin16.6.0 dyld
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... yes
checking how to run the C++ preprocessor... g++ -std=c++11 -E
checking for ld used by g++ -std=c++11... /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld
checking if the linker (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld) is GNU ld... no
checking whether the g++ -std=c++11 linker (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld) supports shared libraries... yes
checking for g++ -std=c++11 option to produce PIC... -fno-common -DPIC
checking if g++ -std=c++11 PIC flag -fno-common -DPIC works... yes
checking if g++ -std=c++11 static flag -static works... no
checking if g++ -std=c++11 supports -c -o file.o... yes
checking if g++ -std=c++11 supports -c -o file.o... (cached) yes
checking whether the g++ -std=c++11 linker (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld) supports shared libraries... yes
checking dynamic linker characteristics... darwin16.6.0 dyld
checking how to hardcode library paths into programs... immediate
checking for ar... /usr/bin/ar
checking for ranlib... /usr/bin/ranlib
>

Could also be something wonky about your paths that was setup for building something else.  My install is pretty fresh, and only setup to build bitcoin and qtum.  Here is my .bash_profile:

> ~/ $ cat .bash_profile
export PATH="/usr/local/opt/berkeley-db@4/bin:$PATH"
export PATH="/usr/local/opt/qt/bin:$PATH"
export PATH="/usr/local/opt/gettext/bin:$PATH"
export PATH="/usr/local/opt/sqlite/bin:$PATH"
export PATH="/usr/local/opt/icu4c/bin:$PATH"
export PATH="/usr/local/opt/icu4c/sbin:$PATH"
export PATH="/usr/local/opt/gnu-sed/libexec/gnubin:$PATH"
MANPATH="/usr/local/opt/gnu-sed/libexec/gnuman:$MANPATH"
>

Iʻm also on the same boost as Neil.
> $ brew info boost
boost: stable 1.64.0 (bottled), HEAD
>

**Comment by rolf-schenk** [comment]
Compared the versions (gcc, clang, ranlib), outputs provided above and also .bash_profile
Only one difference: 
@qtum-neil  has Apple LLVM version 8.0.0 (clang-800.0.42.1)
I have Apple LLVM version 8.1.0 (clang-802.0.42)
@tekahuna also has Apple LLVM version 8.1.0 (clang-802.0.42) 
I am reinstalling Xcode and the command-line tools to be sure it's not gotten mixed up funny for some reasons...

**Comment by qtum-neil** [comment]
@rolf-schenk make sure you uninstall leveldb and gperftools before running autogen.sh as it seems that the problem is related to leveldb. also please post installed versions of the above.

**Comment by rolf-schenk** [comment]
Did brew uninstall of these
- leveldb 1.20
- gperftools 2.5_1
After autogen.sh
- leveldb: stable 1.20 (bottled)
- gperftools: stable 2.5 (bottled), HEAD

- reinstalled XCode Version 8.3.3 (8E3004b)
- checked-out qtum from scratch

Unfortunately backup-wallet via menu is still crashing at my end after fresh build.

I noticed also this warning - seems to be required by eth?

`ld: warning: ignoring file /Library/Frameworks//libcurl.framework/libcurl, missing required architecture x86_64 in file /Library/Frameworks//libcurl.framework/libcurl (2 slices)`

and still seeing this:
`
std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&) in libjsonrpccpp-client.a(httpclient.cpp.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [ethminer/ethminer] Error 1
make[1]: *** [ethminer/CMakeFiles/ethminer.dir/all] Error 2
`
I am currently out of ideas but agree that this looks like a linker setup error at my end?

**Comment by rolf-schenk** [comment]
further info:

brew doctor
`Warning: Some frameworks can be picked up by CMake's build system and likely
cause the build to fail. To compile CMake, you may wish to move these
out of the way:
  /Library/Frameworks/libcurl.framework`

**Comment by tekahuna** [comment]
I donʻt have this framework on my system.  Not sure what you have that might have installed and/or depends on it.  What is the date of file?  Might help you track down what depends on it.

From `qtum/src/cpp-ethereum/cmake/ProjectJsonRpcCpp.cmake`

> \# HTTP client from JSON RPC CPP requires curl library. It can find it itself,
\# but we need to know the libcurl location for static linking.
find_package(CURL REQUIRED)
>

It would seem that in the same way brew picked up this framework, so does cmake during the Ethereum code build process.  

As brew suggests, try moving it out of the way, and rebuilding Qtum and maybe CMake as well.

**Comment by rolf-schenk** [comment]
yep, make sense... now I know what to fix. Thanks! I will try this a bit later.

**Comment by tekahuna** [comment]
Those are all the exact same warnings I get during build.  By fresh wallet, are you saying you recloned the repository and built from scratch again?  If not, you might need to `make clean` in `src/cpp-ethereum`.

Outside of that, I donʻt really get a direction with your stack trace info other than maybe a general system corruption.

Do you have other coin wallets that backup fine?  Did you build them?

Alternatively, and this comes with a huge ***WARNING***, for all the reasons people need to be smart about what they run on their systems they store crypto on.  You **MAY** want to try this build I made for myself to determine if it is your system, or your build system that is the issue:

[https://github.com/tekahuna/qtum/releases/tag/mac_deploy_fix_test](url)

Again, always use discretion when running software from unknown sources.  Although, I swear on Bob Dobbs and all my BOB(Earlz would get the joke), all code changes are found in the release tag, and only related to getting `mac deploy` working.

**Comment by rolf-schenk** [comment]
Thanks @tekahuna - It's also crashing with your distribution at my end. 
I checked the whole mac on hardware side, no issues. 
Now a fresh installation or another mac is in the pipeline for dev - and yes, for dev only. This is really important and thanks for putting up the orange flag once more - can't be said enough times :-)

I am closing this ticket for now.

**Comment by rolf-schenk** [comment]
@tekahuna one thing to add - by fresh wallet I meant this - with qt-qtum, settings, reset config. Then I restarted qt-qtum and I entered a different path for the chain and wallet. Now I have two locations for the chain & wallet = two wallets. 

off-topic:
qtum-cli however seems to pick the wallet from standard path only. Need to check how to set the custom path to be picked by qtumd. I don't see such a command yet in qtum-cli
